diff --git a/pythonFiles/positron/_vendor/jedi_language_server/jedi_utils.py b/pythonFiles/positron/_vendor/jedi_language_server/jedi_utils.py
index 4f3ba7bd5..52524daa2 100644
--- a/pythonFiles/positron/_vendor/jedi_language_server/jedi_utils.py
+++ b/pythonFiles/positron/_vendor/jedi_language_server/jedi_utils.py
@@ -12,9 +12,9 @@ from inspect import Parameter
 from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple, Union
 
 import docstring_to_markdown
-import jedi.api.errors
-import jedi.inference.references
-import jedi.settings
+import jedi
+from jedi.api.errors import SyntaxError as JediSyntaxError
+from jedi import settings
 from jedi import Project, Script
 from jedi.api.classes import (
     BaseName,
@@ -105,14 +105,14 @@ def set_jedi_settings(
     initialization_options: InitializationOptions,
 ) -> None:
     """Sets jedi settings."""
-    jedi.settings.auto_import_modules = list(
+    settings.auto_import_modules = list(
         set(
-            jedi.settings.auto_import_modules
+            settings.auto_import_modules
             + initialization_options.jedi_settings.auto_import_modules
         )
     )
 
-    jedi.settings.case_insensitive_completion = (
+    settings.case_insensitive_completion = (
         initialization_options.jedi_settings.case_insensitive_completion
     )
     if initialization_options.jedi_settings.debug:
@@ -284,7 +284,7 @@ def lsp_document_symbols(names: List[Name]) -> List[DocumentSymbol]:
     return results
 
 
-def lsp_diagnostic(error: jedi.api.errors.SyntaxError) -> Diagnostic:
+def lsp_diagnostic(error: JediSyntaxError) -> Diagnostic:
     """Get LSP Diagnostic from Jedi SyntaxError."""
     return Diagnostic(
         range=Range(
diff --git a/pythonFiles/positron/_vendor/jedi_language_server/__init__.py b/pythonFiles/positron/_vendor/jedi_language_server/__init__.py
index ba6eaf9fe..28266bd95 100644
--- a/pythonFiles/positron/_vendor/jedi_language_server/__init__.py
+++ b/pythonFiles/positron/_vendor/jedi_language_server/__init__.py
@@ -1,4 +1,3 @@
 """Jedi Language Server."""
-from importlib.metadata import version
 
-__version__ = version("jedi-language-server")
+__version__ = "unknown"
